hydrolib/core/io/bc/models.py

```diff
@@ -236,6 +291,152 @@ class T3D(ForcingBase):
+        "timeinterpolation", enum=TimeInterpolation
+    )
+
+    @root_validator(pre=True)
+    def _validate_quantityunitpair(cls, values):
+        super()._validate_quantityunitpair(values)
+
+        quantityunitpairkey = "quantityunitpair"
+        quantityunitpairs = values[quantityunitpairkey]
+
+        T3D._validate_that_first_unit_is_time_and_has_no_verticalpositionindex(
+            quantityunitpairs
+        )
+
+        verticalpositionindexes = values.get("verticalpositionindex")
+        number_of_verticalpositions = len(values["verticalpositions"])
+
+        if verticalpositionindexes is None:
+            T3D._validate_that_all_non_time_quantityunitpairs_have_valid_verticalpositionindex(
+                quantityunitpairs, number_of_verticalpositions
+            )
+            return values
+
+        T3D._validate_verticalpositionindexes_and_update_quantityunitpairs(
+            verticalpositionindexes,
+            number_of_verticalpositions,
+            quantityunitpairs,
+        )
+
+        return values
+
+    @staticmethod
+    def _validate_that_first_unit_is_time_and_has_no_verticalpositionindex(
+        quantityunitpairs: List[QuantityUnitPair],
+    ) -> None:
+        if quantityunitpairs[0].quantity.lower() != "time":
+            raise ValueError("First quantity should be `time`")
+        if quantityunitpairs[0].verticalpositionindex is not None:
+            raise ValueError("`time` quantity cannot have vertical position index")
+
+    @staticmethod
+    def _validate_that_all_non_time_quantityunitpairs_have_valid_verticalpositionindex(
+        quantityunitpairs: List[QuantityUnitPair], maximum_verticalpositionindex: int
+    ) -> None:
+        for quantityunitpair in quantityunitpairs:
+            quantity = quantityunitpair.quantity.lower()
+            verticalpositionindex = quantityunitpair.verticalpositionindex
+
+            if quantity == "time":
+                continue
+            if not T3D._is_valid_verticalpositionindex(
+                verticalpositionindex, maximum_verticalpositionindex
+            ):
+                raise ValueError(
+                    f"Vertical position index should be between 1 and {maximum_verticalpositionindex}"
+                )
+
+    @staticmethod
+    def _validate_verticalpositionindexes_and_update_quantityunitpairs(
+        verticalpositionindexes: List[int],
+        number_of_verticalpositions: int,
+        quantityunitpairs: List[QuantityUnitPair],
+    ) -> None:
+        if verticalpositionindexes is None:
+            raise ValueError("verticalpositionindex is not provided")
+
+        number_of_quantityunitpairs = len(quantityunitpairs)
+        number_of_verticalpositionindexes = len(verticalpositionindexes)
+
+        if number_of_verticalpositionindexes == number_of_quantityunitpairs:
+            T3D._validate_verticalpositionindexes_and_add_them_to_quantityunitpairs(
+                verticalpositionindexes[1:],
```

Comment:
Which case are you handling here? When are these numbers equal, and why are you then skipping element 0?

Reply:
We discussed this via Teams.

The idea here was that the user could specify the `verticalpositionindex` in two ways. The first way would be an array with a length equal to the number of quantities/units given. The second option would be an array with a length equal to the quantities/units given - 1.

In the first case (same length arrays) the user has specified a `verticalpositionindex` for all quantities, including `time`. Since `time` cannot have a `verticalpositionindex`, the code you highlighted skips the first entry of the `verticalpositionindexes`, bascially ignoring the value the user entered.

We discussed that this is not what we want and that I will change it to the following:
The user should always provide an array of `verticalpositionindex` that is 1 less than the number of quantities/units provided by the user. If the user specifies more (or less) values, a `ValueError` will be raised.

So the only valid option for the user would be:
`
quantities = ["time", "randomQuantity1", "randomQuantity2"]
units = ["randomUnit", "randomUnit", "randomUnit"]
verticalpositionindex = [1, 2]
`

Reply:
Fixed!